File: backend/src/auth/service.ts
import type { OidcIdentity } from "./oidc.js";
import { validateIdentity } from "./oidc.js";
import { defaultSessionTtlMs, hashSessionToken, newSessionToken } from "./session.js";

export type { OidcIdentity };

export type AuthStore = {
  findOrCreateUserByIdentity(identity: OidcIdentity): Promise<string>;
  createSession(userId: string, tokenHash: string, expiresAt: Date): Promise<void>;
  findIdentityBySessionHash(tokenHash: string, now: Date): Promise<OidcIdentity | null>;
  revokeSession(tokenHash: string): Promise<void>;
};

export type AuthProvider = {
  authorizationUrl(state: string): string;
  exchange(code: string): Promise<OidcIdentity>;
};

export type AuthService = {
  loginUrl(): Promise<{ redirectUrl: string; state: string }>;
  completeLogin(code: string, state: string): Promise<{
    identity: OidcIdentity;
    sessionToken: string;
    csrfToken: string;
  }>;
  session(sessionToken: string): Promise<OidcIdentity | null>;
  logout(sessionToken: string): Promise<void>;
};

export function createAuthService(options: {
  provider: AuthProvider;
  store: AuthStore;
  now?: () => Date;
}): AuthService {
  const now = options.now ?? (() => new Date());

  return {
    async loginUrl() {
      const state = newSessionToken();
      return { redirectUrl: options.provider.authorizationUrl(state), state };
    },
    async completeLogin(code: string) {
      const identity = await options.provider.exchange(code);
      validateIdentity(identity);
      const userId = await options.store.findOrCreateUserByIdentity(identity);
      const sessionToken = newSessionToken();
      await options.store.createSession(userId, hashSessionToken(sessionToken), new Date(now().getTime() + defaultSessionTtlMs));
      return { identity, sessionToken, csrfToken: newSessionToken() };
    },
    async session(sessionToken: string) {
      return options.store.findIdentityBySessionHash(hashSessionToken(sessionToken), now());
    },
    async logout(sessionToken: string) {
      await options.store.revokeSession(hashSessionToken(sessionToken));
    }
  };
}

export function disabledAuthService(): AuthService {
  return {
    async loginUrl() {
      throw new Error("auth is not configured");
    },
    async completeLogin() {
      throw new Error("auth is not configured");
    },
    async session() {
      return null;
    },
    async logout() {
      return undefined;
    }
  };
}

File: backend/internal/auth/service.go
package auth

import (
	"context"
	"time"
)

const DefaultSessionTTL = 30 * 24 * time.Hour

type Provider interface {
	AuthorizationURL(state string) (string, error)
	Exchange(ctx context.Context, code string) (OIDCIdentity, error)
}

type Store interface {
	FindOrCreateUserByIdentity(ctx context.Context, identity OIDCIdentity) (string, error)
	CreateSession(ctx context.Context, userID string, tokenHash string, expiresAt time.Time) error
	FindIdentityBySessionHash(ctx context.Context, tokenHash string, now time.Time) (OIDCIdentity, bool, error)
	RevokeSession(ctx context.Context, tokenHash string) error
}

type ServiceOptions struct {
	Provider Provider
	Store    Store
	Now      func() time.Time
}

type Service struct {
	provider Provider
	store    Store
	now      func() time.Time
}

func NewService(options ServiceOptions) Service {
	now := options.Now
	if now == nil {
		now = time.Now
	}
	return Service{
		provider: options.Provider,
		store:    options.Store,
		now:      now,
	}
}

func (s Service) LoginURL(ctx context.Context) (string, string, error) {
	state, err := NewSessionToken()
	if err != nil {
		return "", "", err
	}
	loginURL, err := s.provider.AuthorizationURL(state)
	if err != nil {
		return "", "", err
	}
	return loginURL, state, nil
}

func (s Service) CompleteLogin(ctx context.Context, code string, _ string) (OIDCIdentity, string, string, error) {
	identity, err := s.provider.Exchange(ctx, code)
	if err != nil {
		return OIDCIdentity{}, "", "", err
	}
	if err := identity.Validate(); err != nil {
		return OIDCIdentity{}, "", "", err
	}

	userID, err := s.store.FindOrCreateUserByIdentity(ctx, identity)
	if err != nil {
		return OIDCIdentity{}, "", "", err
	}

	sessionToken, err := NewSessionToken()
	if err != nil {
		return OIDCIdentity{}, "", "", err
	}
	tokenHash, err := HashSessionToken(sessionToken)
	if err != nil {
		return OIDCIdentity{}, "", "", err
	}
	expiresAt := s.now().Add(DefaultSessionTTL)
	if err := s.store.CreateSession(ctx, userID, tokenHash, expiresAt); err != nil {
		return OIDCIdentity{}, "", "", err
	}

	csrfToken, err := NewSessionToken()
	if err != nil {
		return OIDCIdentity{}, "", "", err
	}

	return identity, sessionToken, csrfToken, nil
}

func (s Service) Session(ctx context.Context, sessionToken string) (OIDCIdentity, bool) {
	tokenHash, err := HashSessionToken(sessionToken)
	if err != nil {
		return OIDCIdentity{}, false
	}
	identity, ok, err := s.store.FindIdentityBySessionHash(ctx, tokenHash, s.now())
	if err != nil {
		return OIDCIdentity{}, false
	}
	return identity, ok
}

func (s Service) Logout(ctx context.Context, sessionToken string) error {
	tokenHash, err := HashSessionToken(sessionToken)
	if err != nil {
		return err
	}
	return s.store.RevokeSession(ctx, tokenHash)
}
